File: tsm-rucklidge.c
/*
 * Rucklidge Attractor
 *
 * (c) 2018-2025 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
 */
#include <stdio.h>
#include <stdlib.h>
#include <mpfr.h>
#include "taylor-ode.h"

struct Parameters { real alpha, kappa; };

model *tsm_init_p (int argc, char **argv, int n) { (void)n;
    CHECK(argc == 11);
    model *_ = malloc(sizeof (model)); CHECK(_);
    tsm_get_p(argv, argc, &_->alpha, &_->kappa);
    return _;
}

void ode (triplet *v, series x, series y, series z, model *_, int k) {
    //  x' = ay - kx - yz
    mpfr_fmms(v->x, _->alpha, y[k], _->kappa, x[k], RND);
    mpfr_sub(v->x, v->x, *t_mul(y, z, k), RND);
    //  y' = x
    mpfr_set(v->y, x[k], RND);
    //  z' = y^2 - z
    mpfr_sub(v->z, *t_sqr(y, k), z[k], RND);
}

File: tsm-rucklidge.py
#!/usr/bin/env python3
#
#  Example: ./tsm-rucklidge.py 15 10 0.01 15000 1 0 0 6.7 2
#
#  (c) 2018-2020 m4r35n357@gmail.com (Ian Smith), for licencing see the LICENCE file
#

from sys import argv
from collections import namedtuple
from ad import tsm, t_prod, t_sqr, Components


class Parameters(namedtuple('ParametersType', ['alpha', 'kappa'])):
    pass

def get_p(order):
    return Parameters(alpha = float(argv[8]),
                      kappa = float(argv[9]))

def ode(x, y, z, p, i, k):
    return Components(x = p.alpha * y[k] - p.kappa * x[k] - t_prod(y, z, k),
                      y = x[k],
                      z = t_sqr(y, k) - z[k])

tsm(ode, get_p, None)
